src/annotator/integrations/image-text-layer.js

```diff
@@ -1,29 +1,152 @@
 import debounce from 'lodash.debounce';
 
 import { ListenerCollection } from '../../shared/listener-collection';
+import {
+  rectCenter,
+  rectsOverlapHorizontally,
+  rectsOverlapVertically,
+  unionRects,
+} from '../util/geometry';
 
 /**
- * @typedef Rect
- * @prop {number} left
- * @prop {number} right
- * @prop {number} top
- * @prop {number} bottom
+ * @typedef WordBox
+ * @prop {DOMRect[]} glyphs
+ * @prop {string} text
+ * @prop {DOMRect} rect - Bounding rectangle of all glyphs in word
  */
 
 /**
- * @param {Rect|null} a
- * @param {Rect} b
+ * @typedef LineBox
+ * @prop {WordBox[]} words
+ * @prop {DOMRect} rect - Bounding rectangle of all words in line
  */
-function mergeBoxes(a, b) {
-  if (!a) {
-    return b;
+
+/**
+ * @typedef ColumnBox
+ * @prop {LineBox[]} lines
+ * @prop {DOMRect} rect - Bounding rectangle of all lines in column
+ */
+
+/**
+ * Group characters in a page into words, lines and columns.
+ *
+ * The input is assumed to be _roughly_ reading order, more so at the low (word,
+ * line) level. When the input is not in reading order, the output may be
+ * divided into more lines / columns than expected. Downstream code is expected
+ * to tolerate over-segmentation. This function should try to avoid producing
+ * lines or columns that significantly intersect, as this can impair text
+ * selection.
+ *
+ * @param {DOMRect[]} chars
+ * @param {string} text
+ * @return {ColumnBox[]}
+ */
+function analyzeLayout(chars, text) {
```

Comment:
The implementation of `analyzeLayout` could be refined or even completely redone in many ways. The important parts here are to:

1. Add enough structure to support good text selection behavior
2. Create a separation between layout analysis and DOM construction, so that each part can be understood independently. It may also prove useful in future to extract this out into a separate module for ease of testing.